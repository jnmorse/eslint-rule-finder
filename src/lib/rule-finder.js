const { CLIEngine, Linter, ...eslint } = require('eslint');
const path = require('path');
const {
  filter,
  orderBy,
  forEach,
  difference
} = require('lodash');

class RuleFinder {
  constructor(configFile) {
    const cf = path.join(process.cwd(), configFile);

    const cli = new CLIEngine({
      cwd: process.cwd(),
      useEslintrc: false,
      configFile: cf
    });

    const linter = new Linter();

    this.config = cli.getConfigForFile();
    this.eslintRules = linter.getRules();
    this.pluginRules = new Map();
  }

  getPluginRules() {
    const { plugins } = this.config;

    forEach(plugins, (plugin) => {
      // eslint-disable-next-line
      const pluginConfig = require(`eslint-plugin-${plugin}`);

      forEach(pluginConfig.rules, (value, key) => {
        this.eslintRules.set(`${plugin}/${key}`, {
          deprecated: value.meta.deprecated || false,
          category: value.meta.docs && value.meta.docs.category ? value.meta.docs.category : plugin,
          ...value
        });
      });
    });
  }

  diffRules() {
    this.getPluginRules();
    const { eslintRules, config } = this;

    const diff = new Map();

    for (const [key, value] of eslintRules) {
      if (!config.rules.hasOwnProperty(key)) {
        diff.set(key, {
          deprecated: value.meta.deprecated || false,
          category: value.meta.docs.category,
          ...value
        });
      }
    }

    return diff;
  }

  get unused() {
    return this.diffRules();
  }
}

module.exports = RuleFinder;
