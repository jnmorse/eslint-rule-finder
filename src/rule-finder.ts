import { CLIEngine, Rule, Linter } from 'eslint';
import path from 'path';

const pack = require(`${path.join(process.cwd(), 'package.json')}`);

export interface CurrentRuleDefintion {
  config: Linter.RuleLevel | Linter.RuleLevelAndOptions;
  definition: Rule.RuleModule;
}

export class RuleFinder extends CLIEngine {
  public currentRules: Map<string, CurrentRuleDefintion> = new Map();

  constructor(configFile?: string) {
    super({
      useEslintrc: false,
      configFile: configFile || pack.main,
      cwd: process.cwd()
    });

    const rules = this.getRules();
    const { rules: configRules } = this.getConfigForFile(
      configFile || pack.main
    );

    if (configRules) {
      Object.keys(configRules).forEach(rule => {
        const definition = rules.get(rule);

        if (definition) {
          this.currentRules.set(rule, {
            config: configRules[rule],
            definition
          });
        }
      });
    }
  }

  private isDeprecated(rule: Rule.RuleModule): boolean {
    if (rule.meta && rule.meta.deprecated) {
      return true;
    }

    return false;
  }

  get deprecated(): Map<string, CurrentRuleDefintion> {
    const deprecatedRules: Map<string, CurrentRuleDefintion> = new Map();

    this.currentRules.forEach((value, rule) => {
      const { definition } = value;
      if (definition && definition.meta && definition.meta.deprecated) {
        deprecatedRules.set(rule, value);
      }
    });

    return deprecatedRules;
  }
}
