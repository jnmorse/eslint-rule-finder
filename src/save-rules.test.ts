import { resolve } from 'path';
import fs from 'fs';

import { RuleFinder } from './rule-finder';
import saveCurrentRules from './save-current-rules';

const configFile = resolve(__dirname, './test-fixutes/withoutPlugins.js');
const fileName = 'test.md';

describe('saveCurrentRules()', () => {
  let loadConfig: RuleFinder;

  beforeEach(() => {
    loadConfig = new RuleFinder(configFile);
  });

  afterEach(() => {
    fs.unlinkSync(fileName);
  });

  test('should save rules', () => {
    saveCurrentRules(fileName, loadConfig.currentRules);
    const stats = fs.statSync('test.md');
    const fileContents: string = fs.readFileSync(fileName, {
      encoding: 'utf-8'
    });

    expect(stats.isFile()).toBeTruthy();
    expect(stats.size).toBeGreaterThan(0);
    expect(fileContents).toMatchSnapshot();
  });
});
