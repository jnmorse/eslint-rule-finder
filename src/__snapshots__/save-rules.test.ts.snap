// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saveCurrentRules() should save rules 1`] = `
"# Current Rules

Currently Defined Rules


## semi

\`\`\`json
{
  \\"semi\\": [\\"error\\",\\"never\\"]
}
\`\`\`
    
## curly

\`\`\`json
{
  \\"curly\\": [\\"error\\"]
}
\`\`\`
    
## valid-jsdoc

\`\`\`json
{
  \\"valid-jsdoc\\": [\\"off\\"]
}
\`\`\`
    "
`;
